[Core]: Nimbus crashes on updating CEL for the resource which is not present

## Description
If we specify the CEL expression to select resources in the SecurityIntentBinding (SIB) selector and if there is a typo or the resource itself doesn't exist in the cluster then nimbus crashes.

## Steps to reproduce
1. Run nimbus
```shell
make run
```
3. Apply cel-based SI + SIB
```shell
kubectl apply -f  examples/namespaced/cel-multi-si-sib-namespaced.yaml
```
4. Edit SIB selector's CEL for a resource that doesn't exist or try a typo.
E.g., `labels["app"] == "nginx"` to ` labels["app"] == "ngin"`
```shell
kubectl edit sib multiple-sis-nsscoped-binding
```
and then nimbus crashes with following logs:
```shell
...
2024-02-14T15:09:29+05:30       ERROR   No labels matched the CEL expressions, aborting NimbusPolicy creation due to missing keys in labels     {"controller": "securityintentbinding", "controllerGroup": "intent.security.nimbus.com", "controllerKind": "SecurityIntentBinding", "SecurityIntentBinding": {"name":"multiple-sis-nsscoped-binding","namespace":"default"}, "namespace": "default", "name": "multiple-sis-nsscoped-binding", "reconcileID": "01a3e488-1554-485a-ac60-23ce451636e6"}
github.com/5GSEC/nimbus/pkg/processor/policybuilder.BuildNimbusPolicy
        /Users/anurag/workspace/nimbus/pkg/processor/policybuilder/nimbuspolicy_builder.go:76
github.com/5GSEC/nimbus/internal/controller.(*SecurityIntentBindingReconciler).Reconcile
        /Users/anurag/workspace/nimbus/internal/controller/securityintentbinding_controller.go:66
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
        /Users/anurag/.local/share/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:119
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /Users/anurag/.local/share/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:316
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /Users/anurag/.local/share/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /Users/anurag/.local/share/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:227
2024-02-14T15:09:29+05:30       INFO    Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference        {"controller": "securityintentbinding", "controllerGroup": "intent.security.nimbus.com", "controllerKind": "SecurityIntentBinding", "SecurityIntentBinding": {"name":"multiple-sis-nsscoped-binding","namespace":"default"}, "namespace": "default", "name": "multiple-sis-nsscoped-binding", "reconcileID": "01a3e488-1554-485a-ac60-23ce451636e6"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x78 pc=0x101b66904]

goroutine 125 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        /Users/anurag/.local/share/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:116 +0x1a4
panic({0x101f9dc40?, 0x102ef3460?})
        /opt/homebrew/opt/go/libexec/src/runtime/panic.go:914 +0x218
github.com/5GSEC/nimbus/internal/controller.(*SecurityIntentBindingReconciler).Reconcile(0x1400000fd28, {0x1021f81b0, 0x1400020ac60}, {{{0x140003283e0?, 0x5?}, {0x140006e6380?, 0x14000741cf8?}}})
        /Users/anurag/workspace/nimbus/internal/controller/securityintentbinding_controller.go:81 +0x604
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1021fbe40?, {0x1021f81b0?, 0x1400020ac60?}, {{{0x140003283e0?, 0xb?}, {0x140006e6380?, 0x0?}}})
        /Users/anurag/.local/share/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:119 +0x8c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0x140006a0500, {0x1021f81e8, 0x140004dbae0}, {0x102024000?, 0x140000ce6c0?})
        /Users/anurag/.local/share/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:316 +0x2e8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0x140006a0500, {0x1021f81e8, 0x140004dbae0})
        /Users/anurag/.local/share/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:266 +0x16c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        /Users/anurag/.local/share/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:227 +0x74
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 77
        /Users/anurag/.local/share/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:223 +0x43c
exit status 2
make: *** [run] Error 1
```

## Expected behaviour
Nimbus should handle gracefully scenarios where resources referenced in CEL expressions are either missing or contain typos.

--- a/internal/controller/securityintentbinding_controller.go
+++ b/internal/controller/securityintentbinding_controller.go
@@ -54,8 +54,10 @@ func (r *SecurityIntentBindingReconciler) Reconcile(ctx context.Context, req ctr
 	}
 	logger.Info("SecurityIntentBinding found", "SecurityIntentBinding.Name", req.Name, "SecurityIntentBinding.Namespace", req.Namespace)
 
-	if err = r.createOrUpdateNp(ctx, logger, req); err != nil {
+	if np, err := r.createOrUpdateNp(ctx, logger, req); err != nil {
 		return requeueWithError(err)
+	} else if np == nil {
+		return doNotRequeue()
 	}
 
 	if err = r.updateStatus(ctx, logger, req); err != nil {
@@ -101,14 +103,13 @@ func (r *SecurityIntentBindingReconciler) deleteFn(deleteEvent event.DeleteEvent
 	return ownerExists(r.Client, obj)
 }
 
-func (r *SecurityIntentBindingReconciler) createOrUpdateNp(ctx context.Context, logger logr.Logger, req ctrl.Request) error {
-	// Always fetch the latest CRs so that we have the latest state of the CRs on the
+func (r *SecurityIntentBindingReconciler) createOrUpdateNp(ctx context.Context, logger logr.Logger, req ctrl.Request) (*v1.NimbusPolicy, error) { // Always fetch the latest CRs so that we have the latest state of the CRs on the
 	// cluster.
 
 	var sib v1.SecurityIntentBinding
 	if err := r.Get(ctx, req.NamespacedName, &sib); err != nil {
 		logger.Error(err, "failed to fetch SecurityIntentBinding", "SecurityIntentBinding.Name", req.Name, "SecurityIntentBinding.Namespace", req.Namespace)
-		return err
+		return nil, err
 	}
 
 	var np v1.NimbusPolicy
@@ -118,38 +119,42 @@ func (r *SecurityIntentBindingReconciler) createOrUpdateNp(ctx context.Context, 
 			return r.createNp(ctx, logger, sib)
 		}
 		logger.Error(err, "failed to fetch NimbusPolicy", "NimbusPolicy.Name", req.Name, "NimbusPolicy.Namespace", req.Namespace)
-		return err
+		return nil, err
 	}
 	return r.updateNp(ctx, logger, sib)
 }
 
-func (r *SecurityIntentBindingReconciler) createNp(ctx context.Context, logger logr.Logger, sib v1.SecurityIntentBinding) error {
+func (r *SecurityIntentBindingReconciler) createNp(ctx context.Context, logger logr.Logger, sib v1.SecurityIntentBinding) (*v1.NimbusPolicy, error) {
 	nimbusPolicy, err := policybuilder.BuildNimbusPolicy(ctx, logger, r.Client, r.Scheme, sib)
 	// TODO: Improve error handling for CEL
 	if err != nil {
 		// If error is caused due to CEL then we don't retry to build NimbusPolicy.
 		if strings.Contains(err.Error(), "error processing CEL") {
 			logger.Error(err, "failed to build NimbusPolicy")
-			return nil
+			return nil, nil
 		}
 		logger.Error(err, "failed to build NimbusPolicy")
-		return err
+		return nil, err
+	}
+	if nimbusPolicy == nil {
+		logger.Info("Abort NimbusPolicy creation as no labels matched the CEL expressions")
+		return nil, nil
 	}
 
 	if err := r.Create(ctx, nimbusPolicy); err != nil {
 		logger.Error(err, "failed to create NimbusPolicy", "NimbusPolicy.Name", nimbusPolicy.Name, "NimbusPolicy.Namespace", nimbusPolicy.Namespace)
-		return err
+		return nil, err
 	}
 	logger.Info("NimbusPolicy created", "NimbusPolicy.Name", nimbusPolicy.Name, "NimbusPolicy.Namespace", nimbusPolicy.Namespace)
 
-	return nil
+	return nimbusPolicy, nil
 }
 
-func (r *SecurityIntentBindingReconciler) updateNp(ctx context.Context, logger logr.Logger, sib v1.SecurityIntentBinding) error {
+func (r *SecurityIntentBindingReconciler) updateNp(ctx context.Context, logger logr.Logger, sib v1.SecurityIntentBinding) (*v1.NimbusPolicy, error) {
 	var existingNp v1.NimbusPolicy
 	if err := r.Get(ctx, types.NamespacedName{Name: sib.Name, Namespace: sib.Namespace}, &existingNp); err != nil {
 		logger.Error(err, "failed to fetch NimbusPolicy", "NimbusPolicy.Name", sib.Name, "NimbusPolicy.Namespace", sib.Namespace)
-		return err
+		return nil, err
 	}
 
 	nimbusPolicy, err := policybuilder.BuildNimbusPolicy(ctx, logger, r.Client, r.Scheme, sib)
@@ -158,20 +163,24 @@ func (r *SecurityIntentBindingReconciler) updateNp(ctx context.Context, logger l
 		// If error is caused due to CEL then we don't retry to build NimbusPolicy.
 		if strings.Contains(err.Error(), "error processing CEL") {
 			logger.Error(err, "failed to build NimbusPolicy")
-			return nil
+			return nil, nil
 		}
 		logger.Error(err, "failed to build NimbusPolicy")
-		return err
+		return nil, err
+	}
+	if nimbusPolicy == nil {
+		logger.Info("Abort NimbusPolicy creation as no labels matched the CEL expressions")
+		return nil, nil
 	}
 
 	nimbusPolicy.ObjectMeta.ResourceVersion = existingNp.ObjectMeta.ResourceVersion
 	if err := r.Update(ctx, nimbusPolicy); err != nil {
 		logger.Error(err, "failed to configure NimbusPolicy", "NimbusPolicy.Name", nimbusPolicy.Name, "NimbusPolicy.Namespace", nimbusPolicy.Namespace)
-		return err
+		return nil, err
 	}
 	logger.Info("NimbusPolicy configured", "NimbusPolicy.Name", nimbusPolicy.Name, "NimbusPolicy.Namespace", nimbusPolicy.Namespace)
 
-	return nil
+	return nimbusPolicy, nil
 }
 
 func (r *SecurityIntentBindingReconciler) updateStatus(ctx context.Context, logger logr.Logger, req ctrl.Request) error {

--- a/pkg/adapter/nimbus-kubearmor/manager/manager.go
+++ b/pkg/adapter/nimbus-kubearmor/manager/manager.go
@@ -180,7 +180,7 @@ func deleteKsp(ctx context.Context, npName, npNamespace string) {
 func deleteUnnecessaryKsps(ctx context.Context, currentKsps []kubearmorv1.KubeArmorPolicy, namespace string, logger logr.Logger) {
 	var existingKsps kubearmorv1.KubeArmorPolicyList
 	if err := k8sClient.List(ctx, &existingKsps, client.InNamespace(namespace)); err != nil {
-		logger.Error(err, "Failed to list KubeArmorPolicies for cleanup")
+		logger.Error(err, "failed to list KubeArmorPolicies for cleanup")
 		return
 	}
 
@@ -193,9 +193,7 @@ func deleteUnnecessaryKsps(ctx context.Context, currentKsps []kubearmorv1.KubeAr
 		existingKsp := existingKsps.Items[i]
 		if _, needed := currentKspNames[existingKsp.Name]; !needed {
 			if err := k8sClient.Delete(ctx, &existingKsp); err != nil {
-				logger.Error(err, "Failed to delete unnecessary KubeArmorPolicy", "KubeArmorPolicy.Name", existingKsp.Name)
-			} else {
-				logger.Info("Deleted unnecessary KubeArmorPolicy", "KubeArmorPolicy.Name", existingKsp.Name)
+				logger.Error(err, "failed to delete unnecessary KubeArmorPolicy", "KubeArmorPolicy.Name", existingKsp.Name)
 			}
 		}
 	}

--- a/pkg/processor/policybuilder/common.go
+++ b/pkg/processor/policybuilder/common.go
@@ -65,7 +65,7 @@ func ProcessCEL(ctx context.Context, k8sClient client.Client, namespace string, 
 				"labels": resource["labels"],
 			})
 			if err != nil {
-				logger.Error(err, "Error evaluating CEL expression for pod", "PodName", pod.Name)
+				logger.Info("Error evaluating CEL expression for pod", "PodName", pod.Name, "error", err.Error())
 				// Instead of returning an error immediately, we log the error and continue.
 				break
 			}

